common/quotas/global/doc.go

```diff
@@ -0,0 +1,161 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+/*
+Package global contains a global-load-balance-aware ratelimiter (when complete).
+
+# High level overview
+
+At a very high level, this will:
+  - collect usage metrics in-memory in "limiting" hosts (which use an in-memory ratelimiter)
+  - asynchronously submit these usage metrics to "aggregating" hosts, which will return per-key RPS data
+  - "limiting" hosts use this returned value to update their in-memory ratelimiters
+
+"Aggregating" hosts will internally:
+  - receive data for key X from all "limiting" hosts
+  - compute request-weight per host+key in the cluster
+  - return the "limiting"-host's weighted portion of each ratelimit to allow
+
+The exact logic that the "aggregating" hosts use is intentionally hidden from the
+"limiting" hosts, so it can be changed without changing how they enforce limits.
+
+The weight-calculation algorithm will (eventually) be controlled by a dynamicconfig value,
+to allow shadowing and experimenting with different algorithms at runtime, or disabling
+the global logic altogether.
+
+See sub-packages for implementation details.
+
+# Data flow in the system
+
+The overall planned system's data-flow is driven entirely by limiting hosts repeatedly
+calling an "update(request, data)" API on aggregating hosts, and using the return value.
+Aggregating hosts make no outbound requests at all:
+
+	lim: Frontend host limiting e.g. user RPS for domain ASDF
+	agg: History host aggregating this data, to compute fair RPS allowances
+	┌───┐              ┌────┐┌────┐┌────┐              ┌────┐
+	│lim│              │ring││agg1││agg2│              │lim2│
+	└─┬─┘              └─┬──┘└─┬──┘└─┬──┘              └─┬──┘
+	  │                  │     │     │                   │
+	  │ shard(a,b,c)     │     │     │    update(a1,d37) │
+	  │────────────────>┌┴┐   ┌┴┐<───────────────────────│
+	  │    [[a], [b,c]] │ │   └┬┘───────────────────────>│
+	  │<────────────────└┬┘    │     │                  ...
+	  │                  │     │     │
+	  │ update(a1)       │     │     │
+	  │──────────────────────>┌┴┐    │
+	  │                  │    │ │    │
+	  │ update(b1,c2)    │    │ │    │
+	  │────────────────────────────>┌┴┐
+	  │                  │    │ │   │ │
+	  │      [b/4, c3/4] │    │ │   │ │
+	  │<────────────────────────────└┬┘
+	  │                  │    │ │    │
+	  │            [a/2] │    │ │    │
+	  │<──────────────────────└┬┘    │
+	  │                  │     │     │
+	┌─┴─┐              ┌─┴──┐┌─┴──┐┌─┴──┐
+	│lim│              │ring││agg1││agg2│
+	└───┘              └────┘└────┘└────┘
+
```

Comment:
diagram 2: sequence boogaloo.

renders very nicely in goland (yay correct monospace font usage):
<img width="528" alt="Screenshot 2024-02-26 at 14 42 35" src="https://github.com/uber/cadence/assets/77197/6f4e646d-89b3-4dbd-8413-8e487d767bb1">

and godocs looks fine too (we'll have to see how the public site looks, as that's not open source and runnable AFAIK):
![Screenshot 2024-02-26 at 14 42 17](https://github.com/uber/cadence/assets/77197/72413077-e791-495c-9548-42fd2082c14b)

and of course vscode is useful as ever:
![Screenshot 2024-02-26 at 14 51 41](https://github.com/uber/cadence/assets/77197/3bd665d0-ebc8-4e0b-817e-6f8e658f0ad9)

though if you choose a different-but-same-thing, you can get package docs from vscode:
![Screenshot 2024-02-26 at 14 51 59](https://github.com/uber/cadence/assets/77197/b83a9279-d7ee-4224-a9e9-18f2a98f752a)